File: python/src/threads/store/inbox.py
"""The durable channel inbox.

A verified batch is inserted whole in one transaction, each item under its own key, before the
webhook is answered. A run consumes an item by appending its event with `consume` bound to the
append, which sets `consumed_seq` in the same transaction.
"""

import sqlite3
from collections.abc import Callable, Sequence
from dataclasses import dataclass

from threads.log import ParseError, ThreadId
from threads.store.companion import Companion
from threads.store.sql import blob_of, int_of, text_of, transaction
from threads.store.verify import StoredEvent


@dataclass(frozen=True, slots=True)
class Item:
    """One parsed item of a verified batch, keyed for dedup and addressed to a conversation."""

    channel: str
    installation_id: str
    item_key: str
    delivery_id: str
    address: str
    item: bytes
    """The Inbound item's canonical JSON."""


@dataclass(frozen=True, slots=True)
class Row:
    inbox_id: int
    channel: str
    installation_id: str
    item_key: str
    delivery_id: str
    thread_id: ThreadId
    item: bytes


def insert_batch(
    conn: sqlite3.Connection,
    tenant_id: str,
    items: Sequence[Item],
    now: int,
    new_thread: Callable[[], ThreadId],
) -> frozenset[ThreadId]:
    """Every item in one transaction; a duplicate is a no-op. The address maps to its thread by
    an atomic create-or-get. Returns the threads that have unconsumed items now."""
    threads: set[ThreadId] = set()
    with transaction(conn):
        for item in items:
            thread = _thread(conn, tenant_id, item, new_thread)
            conn.execute(
                "INSERT INTO inbox (tenant_id, channel, installation_id, item_key, delivery_id,"
                " thread_id, item, received_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)"
                " ON CONFLICT DO NOTHING",
                (
                    tenant_id,
                    item.channel,
                    item.installation_id,
                    item.item_key,
                    item.delivery_id,
                    thread,
                    item.item,
                    now,
                ),
            )
            threads.add(thread)
    return frozenset(threads)


def _thread(
    conn: sqlite3.Connection, tenant_id: str, item: Item, new_thread: Callable[[], ThreadId]
) -> ThreadId:
    where = (tenant_id, item.channel, item.installation_id, item.address)
    conn.execute(
        "INSERT INTO channel_threads (tenant_id, channel, installation_id, address, thread_id)"
        " VALUES (?, ?, ?, ?, ?) ON CONFLICT DO NOTHING",
        (*where, new_thread()),
    )
    (thread,) = conn.execute(
        "SELECT thread_id FROM channel_threads WHERE tenant_id = ? AND channel = ?"
        " AND installation_id = ? AND address = ?",
        where,
    ).fetchone()
    return ThreadId(text_of(thread))


def pending(conn: sqlite3.Connection, tenant_id: str, thread_id: ThreadId) -> tuple[Row, ...]:
    """The thread's unconsumed items, in arrival order."""
    return _rows(
        conn,
        "WHERE tenant_id = ? AND thread_id = ? AND consumed_seq IS NULL ORDER BY inbox_id",
        (tenant_id, thread_id),
    )


def all_rows(conn: sqlite3.Connection, tenant_id: str) -> tuple[Row, ...]:
    """Every row of the tenant, in insertion order."""
    return _rows(conn, "WHERE tenant_id = ? ORDER BY inbox_id", (tenant_id,))


def unconsumed_threads(conn: sqlite3.Connection) -> tuple[tuple[str, ThreadId], ...]:
    """(tenant, thread) of every thread with an unconsumed item: what a restarted host drains."""
    rows: list[tuple[object, object]] = conn.execute(
        "SELECT DISTINCT tenant_id, thread_id FROM inbox WHERE consumed_seq IS NULL"
    ).fetchall()
    return tuple((text_of(t), ThreadId(text_of(th))) for t, th in rows)


def _rows(conn: sqlite3.Connection, where: str, args: tuple[str, ...]) -> tuple[Row, ...]:
    found: list[tuple[object, ...]] = conn.execute(
        # Fixed WHERE fragments from this module; every value is bound.
        "SELECT inbox_id, channel, installation_id, item_key, delivery_id, thread_id, item"  # noqa: S608
        " FROM inbox " + where,
        args,
    ).fetchall()
    return tuple(
        Row(
            int_of(i),
            text_of(c),
            text_of(inst),
            text_of(k),
            text_of(d),
            ThreadId(text_of(t)),
            blob_of(item),
        )
        for i, c, inst, k, d, t, item in found
    )


@dataclass(frozen=True, slots=True)
class Conversation:
    channel: str
    installation_id: str
    address: str


def conversation(
    conn: sqlite3.Connection, tenant_id: str, thread_id: ThreadId
) -> Conversation | None:
    """The conversation a channel thread answers to, or None for a thread no channel owns."""
    row: tuple[object, object, object] | None = conn.execute(
        "SELECT channel, installation_id, address FROM channel_threads"
        " WHERE tenant_id = ? AND thread_id = ?",
        (tenant_id, thread_id),
    ).fetchone()
    return None if row is None else Conversation(*(text_of(v) for v in row))


def consume(inbox_id: int) -> Companion:
    """Marks the item consumed by the append's first event, once."""

    def mark(conn: sqlite3.Connection, events: Sequence[StoredEvent]) -> ParseError | None:
        done = conn.execute(
            "UPDATE inbox SET consumed_seq = ? WHERE inbox_id = ? AND consumed_seq IS NULL",
            (events[0].seq, inbox_id),
        )
        if done.rowcount == 1:
            return None
        return ParseError("seq_conflict", f"inbox item {inbox_id} is already consumed")

    return mark

File: typescript/packages/host/src/inbox.ts
import { Inbound } from "@threads/core";
import {
  type Alongside,
  Int,
  type LogError,
  parseRows,
  type Result,
  reading,
  type Sql,
  ThreadId,
  type Tx,
} from "@threads/core/host";
import { z } from "zod";

// store.sql inbox and channel_threads: the durable intake a run consumes from under the branch
// lease.

export type Conversation = {
  readonly tenant: string;
  readonly channel: string;
  readonly installation: string;
  readonly address: string;
};

const ThreadRow: z.ZodType<{ readonly thread_id: ThreadId }> = z.strictObject({
  thread_id: ThreadId,
});

/** The conversation's thread: an atomic create-or-get, keyed by the verified identity only. */
export async function threadFor(
  tx: Tx,
  at: Conversation,
  fresh: string,
): Promise<ThreadId> {
  await tx.run(
    `INSERT INTO channel_threads (tenant_id, channel, installation_id, address, thread_id)
      VALUES (?, ?, ?, ?, ?) ON CONFLICT DO NOTHING`,
    [at.tenant, at.channel, at.installation, at.address, fresh],
  );
  const rows = parseRows(
    ThreadRow,
    await tx.all(
      `SELECT thread_id FROM channel_threads WHERE tenant_id = ? AND channel = ?
        AND installation_id = ? AND address = ?`,
      [at.tenant, at.channel, at.installation, at.address],
    ),
  );
  const row = rows.ok ? rows.value[0] : undefined;
  if (row === undefined)
    throw new Error("channel_threads holds the row just written");
  return row.thread_id;
}

export type InboxItem = {
  readonly inbox_id: number;
  readonly channel: string;
  readonly installation_id: string;
  readonly delivery_id: string;
  readonly received_at: number;
  readonly item: Exclude<Inbound, { kind: "ignore" }>;
};

const Row = z.strictObject({
  inbox_id: Int,
  channel: z.string(),
  installation_id: z.string(),
  delivery_id: z.string(),
  received_at: Int,
  item: z.instanceof(Uint8Array),
});

/** The thread's unconsumed items, oldest first. Stored bytes are parsed again (a boundary). */
export async function pendingItems(
  sql: Sql,
  tenant: string,
  threadId: string,
): Promise<Result<readonly InboxItem[], LogError>> {
  const rows = parseRows(
    Row,
    await reading(sql, (tx) =>
      tx.all(
        `SELECT inbox_id, channel, installation_id, delivery_id, received_at, item FROM inbox
          WHERE tenant_id = ? AND thread_id = ? AND consumed_seq IS NULL ORDER BY inbox_id`,
        [tenant, threadId],
      ),
    ),
  );
  if (!rows.ok) return rows;
  const items: InboxItem[] = [];
  for (const { item, ...row } of rows.value) {
    const parsed = Inbound.safeParse(
      JSON.parse(new TextDecoder().decode(item)),
    );
    if (!parsed.success || parsed.data.kind === "ignore")
      return {
        ok: false,
        error: {
          code: "log_corrupt",
          message: "an inbox row fails its schema",
        },
      };
    items.push({ ...row, item: parsed.data });
  }
  return { ok: true, value: items };
}

/**
 * Marks an item consumed; `seq` is the event that recorded it, 0 when none was appended.
 * Conditional on `consumed_seq IS NULL`, so doing it again after an unknown commit is a no-op.
 */
export async function consumed(
  tx: Tx,
  inboxId: number,
  seq: number,
): Promise<boolean> {
  const changed = await tx.run(
    "UPDATE inbox SET consumed_seq = ? WHERE inbox_id = ? AND consumed_seq IS NULL",
    [seq, inboxId],
  );
  return changed === 1;
}

/**
 * The append that applies an item also consumes it, once: when another process applied it
 * first, this append rolls back, so an item never runs twice (found by the F9.6 drill). It is
 * consumed at its channel_delivery's seq (store.sql inbox), or, for a decision or control item,
 * at its append's first event.
 */
export function consumes(inboxId: number): Alongside {
  return async (added, tx) => {
    const applied =
      added.find(
        (l) => l.kind === "event" && l.event.type === "channel_delivery",
      ) ?? added[0];
    if (
      applied?.kind === "event" &&
      (await consumed(tx, inboxId, applied.event.seq))
    )
      return { ok: true, value: undefined };
    return {
      ok: false,
      error: {
        code: "seq_conflict",
        message: `inbox item ${inboxId} is already consumed`,
      },
    };
  };
}

/** Every channel thread, for the replies a restart may owe (spec/schema/README.md). */
export async function channelThreads(
  sql: Sql,
): Promise<
  readonly { readonly tenant_id: string; readonly thread_id: ThreadId }[]
> {
  const rows = parseRows(
    z.strictObject({ tenant_id: z.string(), thread_id: ThreadId }),
    await reading(sql, (tx) =>
      tx.all("SELECT tenant_id, thread_id FROM channel_threads"),
    ),
  );
  return rows.ok ? rows.value : [];
}

/** The conversation a thread was created for, if it is a channel thread. */
export async function conversationOf(
  sql: Sql,
  tenant: string,
  threadId: string,
): Promise<Conversation | undefined> {
  const rows = parseRows(
    z.strictObject({
      channel: z.string(),
      installation_id: z.string(),
      address: z.string(),
    }),
    await reading(sql, (tx) =>
      tx.all(
        `SELECT channel, installation_id, address FROM channel_threads
          WHERE tenant_id = ? AND thread_id = ?`,
        [tenant, threadId],
      ),
    ),
  );
  const row = rows.ok ? rows.value[0] : undefined;
  return row === undefined
    ? undefined
    : {
        tenant,
        channel: row.channel,
        installation: row.installation_id,
        address: row.address,
      };
}
